ref(stats): Introduce discard outcome & reason descriptions (#10585)
getsentry/sentry-docs: docs/product/stats/index.mdx

| Invalid Multipart               | Parsing a multipart form-data request failed.                                                                           |
| Invalid Replay Event            | Sentry failed to parse the replay so we dropped it.                                                                     |
| Invalid Replay Event No Payload | Sentry failed to parse the replay so we dropped it.                                                                     |
| Invalid Replay Event Pii        | Sentry failed to parse the replay so we dropped it.                                                                     |
| Invalid Replay Record Event     | Sentry failed to parse the replay so we dropped it.                                                                     |
| Invalid Replay Video Event      | Sentry failed to parse the replay so we dropped it.                                                                     |
| Invalid Span                    | A span is not valid after normalization.                                                                                |

> | Invalid Transaction             | The event is parseable but semantically invalid. This should only happen with transaction events.                       |

Author: Thank you for your feedback, but we decided to not expose these reasons like this anymore. I have updated the Pull Request 😉

| Timestamp                       | The timestamp of an event was required for processing and either missing out of the supported time range for ingestion. |
| Too Large                       | The event payload exceeds the maximum size limit for the respective endpoint.                                           |
| Transaction Sampled             | The envelope, which contained only a transaction, was dropped by the dynamic sampling rules.                            |

On this page, you can track projects that have experienced recent spikes in activity or are particularly active, which may require closer attention. Additionally, you have the flexibility to adjust the displayed date range, enabling you to narrow down to a specific period or zoom out for a broader overview.

While primarily designed for organization admins, this page is accessible to all team members. This inclusivity is beneficial, especially when non-admin team members need to review these statistics. It allows them to investigate why events associated with their specific projects might be discarded, dropped or filtered.

> ![Overview of Usage Stats page](./img/usage-stats-page.png)

Author: yes it makes sense. Thank you @lizokm! I will update the screenshot

## Usage Stats

The "Usage" tab provides an overview of the events (errors and transactions) and attachments that Sentry has received across your entire organization. It includes a detailed breakdown of each project's consumption categorized as _Accepted_, _Filtered_, _Discarded_, or _Dropped_.

#### Accepted

Events and attachments that were successfully processed and stored.

#### Discarded

> Events and attachments discarded on the client side due to the sampling configuration in the SDK or other factors.

Reviewer: "or other factors" is a bit unspecific. I would either remove this part of the sentence or list more reasons. They are documented here under "`discarded_events`": https://develop.sentry.dev/sdk/client-reports/#envelope-item-payload

#### Filtered

| Reason                                                                              | Meaning                                                                            |
| ----------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------- |
| [Browser Extensions](/concepts/data-management/filtering/#browser-extension-errors) | Filtered by browser extension.                                                     |

> | Denied Name                                                                         | Filtered due to name being denied.                                                 |
> | Disabled Namespace                                                                  | Filtered due to the namespace being disabled.                                      |

Reviewer: These two are related to metrics and will be removed, so let's not list them here.
```suggestion
```

| Invalid CSP                                                                         | Filtered due to an invalid Content Security Policy (CSP) policy.                   |
| [Legacy Browsers](/concepts/data-management/filtering/#legacy-browser-filters)      | Filtered out certain legacy versions of browsers that are known to cause problems. |
| Localhost                                                                           | Filtered due to localhost restriction.                                             |
| Release Version                                                                     | Filtered by release name (version).                                                |
| [Web Crawlers](/concepts/data-management/filtering/#web-crawler-errors)             | Identified as known web crawler.                                                   |
| ? (Others)                                                                          | Filtered due to other generic filtering rules not specified above.                 |

#### Dropped

> Events and attachments discarded by the system due to rate limits, quotas, spike protection, [size limits](/concepts/data-management/size-limits), or invalid data. You can see the common reasons below:

Reviewer: This combines the two outcomes `RATE_LIMITED` and `INVALID`. Do you really want to mix those?

Author: yes, because we are combining them in the code https://github.com/getsentry/sentry/blob/566f4a0a04d9a9dd6f1a56d13808ea715b913e8b/static/app/views/organizationStats/usageStatsOrg.tsx#L391-L396

> | Reason                          | Meaning                                                                                                                 |

Author: I was uncertain about adding this table here since it can get quite lengthy. However, I didn't want to place it elsewhere, as that might make it feel disconnected from the usage stats section. Thoughts?

Reviewer: I would not list those at all. Those outcome reasons are not stable and meant for internal use. Apart from that, they refer to internals of the system that are confusing to a user.

Author: I understand those outcome reasons are for internal use and may confuse users. However, if users can see this info, we should provide at least a brief explanation to ensure clarity. Don't you agree?

Reviewer: Let's expose these reasons in a simplified and grouped way that makes it easier for users to comprehend. I will post an update here with mapped reasons and improved descriptions.

Reviewer: Here is a list that groups all discard reasons so that they can be understood by users. I've also added more relatable descriptions that point at docs to make this more actionable. 

https://gist.github.com/jan-auer/2d29e2ee6334bbcf32514b12b42b31ea

As for the different kinds of outcomes, please show them separately in the following combination at least in charts:

- Accepted (0)
- Filtered (1)
- Rate limited (2) + Abuse (4) + Cardinality Limited (6) -- currently shown as "Dropped" in the UI
- Invalid (3)
- Client Discard (5)

| Project State Pii               | A project state returned by the upstream contained datascrubbing settings that could not be converted to PII config.    |
| Rate Limited                    | Exceeded allowed event rate.                                                                                            |
| Spike Protection                | Activated to protect from any sudden quota-depleting spike.                                                             |
| Security Report                 | Triggered to prevent any sudden spikes that could deplete the quota.                                                    |
| Security Report Time            | The security report was not recognized due to missing data.                                                             |
| Timestamp                       | The timestamp of an event was required for processing and either missing out of the supported time range for ingestion. |
| Too Large                       | The event payload exceeds the maximum size limit for the respective endpoint.                                           |

> | Transaction Sampled             | The envelope, which contained only a transaction, was dropped by the dynamic sampling rules.                            |

Author: @jan-auer, could you please confirm that the information listed here and in the table above is correct?

  event type.
</Note>

There are three tabs available on the page: [Usage](#usage-stats), [Issues](#issue-stats), and [Health](#health-stats).

The "Usage" tab shows your organization’s event and attachments usage, while also providing a high-level overview of how different projects are using Sentry. The "Issues" tab shows a breakdown of the issues assigned to your team and how they triaged them, while the "Health" tab provides an overview of your team’s project health.

## Usage Stats

The "Usage" tab provides an overview of the events (errors and transactions) and attachments that Sentry has received across your entire organization. It includes a detailed breakdown of each project's consumption categorized as _Accepted_, _Filtered_, _Rate Limited_, _Invalid_ or _Client Discard_.

### Accepted

Events and attachments that were successfully processed and stored.

### Filtered

Filtered events and attachments intentionally excluded based on definied settings. The following reasons are currently defined:

- **Browser Extensions**: Filtered by browser extension.
- **Error Message**: Filtered based on specific error messages.

> - **Filtered Transactions**: Filtered because it was a call to a filtered transaction.

Author: ```suggestion
- **Filtered Transaction**: Filtered because it was a call to a filtered transaction.
```

> - **Browser Extensions**: Filtered by browser extension.
> - **Error Message**: Filtered based on specific error messages.
> - **Filtered Transaction**: Filtered because it was a call to a filtered transaction.
> - **Invalid CSP**: Filtered due to an invalid Content Security Policy (CSP) policy.
> - **Ip Address**: Filtered by IP address.
> - **Legacy Browsers**: Filtered by certain legacy versions of browsers that are known to cause problems.
> - **Localhost**: Filtered due to localhost restriction.
> - **Release Version**: Filtered by release name (version).
> - **Web Crawlers**: Identified as a known web crawler.
> - **Other**: Filtered due to other generic filtering rules not specified above.

Author: ```suggestion
- **Browser Extensions**: Filtered by browser extension.
- **Error Message**: Filtered based on specific error messages.
- **Filtered Transaction**: Filtered because it was a call to a filtered transaction.
- **Invalid CSP**: Filtered due to an invalid Content Security Policy (CSP) policy.
- **Ip Address**: Filtered by IP address.
- **Legacy Browsers**: Filtered by certain legacy versions of browsers that are known to cause problems.
- **Localhost**: Filtered due to localhost restriction.
- **Release Version**: Filtered by release name (version).
- **Web Crawlers**: Identified as a known web crawler.
- **Other**: Filtered due to other generic filtering rules not specified above.
```

For more details, please consult the [Inbound Filters](/concepts/data-management/filtering/) documentation.

### Rate Limited
Events and attachments discarded due to rate limits or quota. The following reasons are currently defined:

> - **Quota**: The monthly quota of your subscription was depleted and there is no on-demand budget left. See [Adjusting your Quota](/pricing/quotas/#adjusting-your-quota) for more information.

Reviewer 2: replace "on-demand" with "pay-as-you-go", as we're updating docs to reflect the latest terminology based on AM3

- **Cardinality Limited**: Exceeded internally defined cardinality limit.
- **Rate Limited**: Exceeded allowed event rate.

> - **Spike Protection**: Activated to protect from any sudden quota-depleting spike. See [Spike Protection](/pricing/quotas/spike-protection/) for more information.

Author: @jan-auer what is your opinion on this list? I might remove "Rate Limited" as it can be confusing

Reviewer: Here's a suggestion for how to structure rate limits. Please expand this with links to relevant documentation as appropriate:

```md
- **Quota**: The monthly quota of your subscription was depleted and there is no on-demand budget left.
- **Key Limit**: Traffic exceeded a rate limit defined on the client key (DSN).
- **Spike Protection**: Activated to protect from a sudden spike in data volume. See [Spike Protection](/pricing/quotas/spike-protection/) for more information.
- **Internal Limit**: A rate limit for excessive volume was enforced by Sentry. These limits are not configurable.
```

Reason codes for the above are:
- quota: `*_usage_exceeded` (everything that ends with `"_usage_exceeded"`)
- key limit: `key_quota`
- spike protection: `smart_rate_limit`
- internal: _(everything else)_

Therefore, please include the "abuse" and "cardinality" limit outcomes under "Internal Limits" for now.

> - **Spike Protection**: Activated to protect from a sudden spike in data volume. See [Spike Protection](/pricing/quotas/spike-protection/) for more information.

Reviewer 2: suggestion: "event volume" instead of "data volume"?

Author: @jan-auer what do you think?

Reviewer: I originally suggested "data" instead of "event", since spike protection could also apply to non-events. Today, that's not the case, so let's go with "event volume" 👍

> - **Spike Protection**: Activated to protect from a sudden spike in data volume. See [Spike Protection](/pricing/quotas/spike-protection/) for more information.

Author: ```suggestion
- **Spike Protection**: Activated to protect from a sudden spike in event volume. See [Spike Protection](/pricing/quotas/spike-protection/) for more information.
```

### Invalid

> Events and attachments discarded due to non-conformity to the expected schema or unprocessable data. The following reasons are currently defined:

Author: ```suggestion
Events and attachments might be discarded if they don't match the expected format or contain data that can't be processed. The following reasons are currently defined:
```

- **Cors**: A request from an unallowed origin was sent to Sentry. Check project settings to configure allowed domains.
- **Duplicate**: An event with the same id has already been processed for this project. Sentry does not allow duplicate events and only stores the first one.
- **Empty**: An empty request or a request missing its primary data component was submitted.
- **Invalid Data**: Data sent by the SDK was invalid and does not meet the basic schema. This likely indicates an SDK implementation bug. Please report a bug if this issue persists.
- **Invalid Request**: The HTTP submission was invalid or was missing required information. This is likely due to a problem in the SDK implementation. Please report a bug if this issue persists.
- **Minidump**: The minidump in a minidump upload request was missing or invalid. See the [Minidumps documentation](/platforms/native/guides/minidumps/) for details on how to upload minidumps.

> - **Other**: An internal problem at Sentry prevented regular processing or storage of the submitted information.

Reviewer: Let's move "other" to the very bottom. 

I'm wondering if it would become more clear if we call this "internal", what do you think?

Reviewer: Not "invalid", but "internal"

Author: oh sorry....misread it - internal sounds good

- **Internal**: An internal problem at Sentry prevented regular processing or storage of the submitted information.

### Client Discard
Events and attachments discarded by the SDK. The following reasons are currently defined:

- **Queue Overflow**: An SDK internal queue (for example, a transport queue) overflowed.
- **Cache Overflow**: An SDK internal cache (for example, an offline event cache) overflowed.
- **Ratelimit Backoff**: The SDK dropped events because a previous rate limit instructed the SDK to back off.
- **Network Error**: Events were dropped due to network errors, and were not retried.
- **Sample Rate**: An event was dropped because of the configured sample rate.
- **Before Send**: An event was dropped in `before_send`.
- **Event Processor**: An event was dropped by an event processor; this may also be used for ignored exception/errors.
- **Send Error**: An event was dropped because of an error when sending it (for example, a 400 response).

> - **Internal Sdk Error**: An event was dropped due to an internal SDK error (for example, a web worker crash).

Reviewer 2: capitalize SDK
